src/Stream.cpp

```diff
@@ -79,46 +75,44 @@ int Stream::converter_add(InstanceSet &kset, Converter::sptr conv,
 
 void Stream::error_in_epics() { epics_client->error_in_epics(); }
 
-int32_t
-Stream::fill_conversion_work(Ring<std::unique_ptr<ConversionWorkPacket>> &q2,
-                             uint32_t max,
-                             std::function<void(uint64_t)> on_seq_data) {
+int32_t Stream::fill_conversion_work(
+    moodycamel::ConcurrentQueue<std::unique_ptr<ConversionWorkPacket>> &q2,
+    uint32_t max, std::function<void(uint64_t)> on_seq_data) {
   uint32_t n0 = 0;
   uint32_t n1 = 0;
-  ulock l1(emit_queue->mx);
-  ulock l2(q2.mx);
-  uint32_t n2 = emit_queue->size_unsafe();
-  uint32_t n3 = (std::min)(max, q2.capacity_unsafe() - q2.size_unsafe());
+  uint32_t n2 = emit_queue->size_approx();
+  uint32_t n3 = max;
   uint32_t ncp = conversion_paths.size();
   std::vector<ConversionWorkPacket *> cwp_last(conversion_paths.size());
   while (n0 < n2 && n3 - n1 >= ncp) {
-    auto e = emit_queue->pop_unsafe();
+    std::unique_ptr<FlatBufs::EpicsPVUpdate> e;
+    auto found = emit_queue->try_dequeue(e);
     n0 += 1;
-    if (e.first != 0) {
+    if (!found) {
       CLOG(6, 1, "empty? should not happen");
       break;
     }
-    auto &up = e.second;
+    auto &up = e;
     if (!up) {
       LOG(6, "empty epics update");
       continue;
     }
     size_t cpid = 0;
     uint32_t ncp = conversion_paths.size();
-    on_seq_data(e.second->seq_data);
+    on_seq_data(e->seq_data);
     for (auto &cp : conversion_paths) {
       auto p = std::unique_ptr<ConversionWorkPacket>(new ConversionWorkPacket);
       cwp_last[cpid] = p.get();
       p->cp = cp.get();
       if (ncp == 1) {
         // more common case
-        p->up = std::move(e.second);
+        p->up = std::move(e);
       } else {
         p->up = std::unique_ptr<FlatBufs::EpicsPVUpdate>(
-            new FlatBufs::EpicsPVUpdate(*e.second));
+            new FlatBufs::EpicsPVUpdate(*e));
       }
-      auto x = q2.push_unsafe(p);
-      if (x != 0) {
+      bool x = q2.enqueue(std::move(p));
+      if (x) {
         CLOG(6, 1, "full? should not happen");
         break;
```

Comment:
Can this log message (and the ones above) have more detail.
E.g. `Conversion work queue is full`